workers/unity/Packages/com.improbable.gdk.core/Subscriptions/TaskManagers/CommandTaskManager.cs

```diff
@@ -0,0 +1,46 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Improbable.Gdk.Core;
+using Improbable.Gdk.Core.Commands;
+using Unity.Entities;
+
+namespace Improbable.Gdk.Subscriptions
+{
+    public class CommandTaskManager<T> : ITaskManager where T : IReceivedCommandResponse
+    {
+        private readonly Dictionary<long, TaskCompletionSource<T>> registeredTasks =
+            new Dictionary<long, TaskCompletionSource<T>>();
+
+        private readonly CommandSystem commandSystem;
+
+        public CommandTaskManager(World world)
+        {
+            commandSystem = world.GetExistingManager<CommandSystem>();
+        }
+
+        public void RegisterTask(long requestId, TaskCompletionSource<T> task)
+        {
+            if (registeredTasks.ContainsKey(requestId))
+            {
+                task.SetException(new ArgumentException("Request Id is already registered."));
+                return;
+            }
+            registeredTasks.Add(requestId, task);
+        }
+
+        public void CompleteTasks()
+        {
+            var responses = commandSystem.GetResponses<T>();
+            foreach (var response in responses)
+            {
+                var requestId = response.GetRequestId();
+                if (registeredTasks.TryGetValue(requestId, out var task))
+                {
+                    task.SetResult(response);
```

Comment:
I think I'd expect this to set and exception in the case the command failed. Then we would probably have to have the reason and context as members of the exception.

Reply:
hm... I don't really agree with that. If we set this as an exception, it would mean that we except failed commands to throw instead of simply checking the status code. Seeing as even a bad network connection / timeout could result in a failed command, I don't think it's a good idea to throw every time. It is expected behaviour after all